Storyshots: Triconfig support

ndelangen opened the pull request:
Issue: #8597 

## What I did
Added support for storyshots

ndelangen committed:
Merge branch 'next' into core/magic-entry-storyshots

ndelangen committed:
CHANGE html to use new triconfig

ndelangen committed:
ADD compatibility for triconfig to storyshots

tmeasday commented:
@ndelangen can you target this at your other branch until it is merged? Would make it easier to review

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

ndelangen commented:
@tmeasday done

shilman reviewed (commented):
@ndelangen Is there any code that can be shared between the CSF loading code added here in storyshots and the CSF loading code in `start.js`? Seems similar at first glance.

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/vue/loader.ts:
```diff
@@ -19,7 +19,7 @@ function load(options: StoryshotsOptions) {
   const storybook = require.requireActual('@storybook/vue');
 
-  configure({ configPath, config, storybook });
+  configure({ configPath, ...config, storybook });
```
Can be `configure({ ...options, storybook });` as for other framework?

ndelangen replied:
✅

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/configure.ts:
```diff
@@ -1,27 +1,143 @@
+function getConfigPathParts(input: string): Output {
+  const configDir = path.resolve(input);
+
+  if (fs.lstatSync(configDir).isDirectory()) {
```
I would have inverted `if` statement to directly `return { files: [configDir], stories: [] };` for this edge case.
And then reduce nesting for the nominal/interesting case.

ndelangen replied:
I'm not sure when configDir would be anything else, and I don't know what else it could be. So I wanted to stay on the safe side and really ensure it's a valid directory.

gaetanmaisse commented on addons/storyshots/storyshots-core/src/frameworks/configure.ts:
```diff
@@ -1,27 +1,143 @@
-  const { configPath = '.storybook', config, storybook } = options;
+function configure(
+  options: {
+    storybook: any;
```
Is `storybook` type `ClientApi`?

ndelangen replied:
yes!

ndelangen replied:
✅

tmeasday commented on addons/storyshots/storyshots-core/src/frameworks/configure.ts:
```diff
@@ -1,27 +1,143 @@
+      output.stories = result;
+    }
+    if (isFile(config)) {
+      output.files.push(config);
```
Should we allow both config + preview here? (ie should this be an error?)

ndelangen replied:
done

tmeasday commented on addons/storyshots/storyshots-core/src/frameworks/configure.ts:
```diff
@@ -1,27 +1,143 @@
+      output.files.push(preview);
+    }
+    if (isFile(main)) {
+      const { stories = [] } = require.requireActual(main);
```
We'd need to ensure the preview/config file was loaded *before* we do this require -- otherwise things like decorators might not work properly.

ndelangen replied:
I already do, the stories are a list of filepaths here, that are required further down in the file, AFTER the preview.js is loaded.

tmeasday replied:
Right, sorry misread that.

tmeasday reviewed (changes requested):
> @ndelangen Is there any code that can be shared between the CSF loading code added here in storyshots and the CSF loading code in start.js? Seems similar at first glance.

Couldn't we simply call `configure()` as we do in the other PR?

ndelangen committed:
FIX review comments, CHANGE to only support a single preview.*

ndelangen commented:
> Couldn't we simply call configure() as we do in the other PR?

No, I tried, it won't allow being called twice, and if I ignore that, it breaks.

Long term that code should be extracted out of start.js anyway.
When we do, I'll de-duplicate.

tmeasday commented:
> No, I tried, it won't allow being called twice, and if I ignore that, it breaks.

Why would it be called twice? Is it getting called from preview/config.js? Isn't this going to be a problem with the webpack build too?

shilman commented:
cc @igor-dv @Hypnosphi

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

# Conflicts:
#	addons/storyshots/storyshots-core/package.json
#	examples/html-kitchen-sink/.storybook/presets.js

ndelangen committed:
FIX bug with resolving of preview.js (doh)

ndelangen commented:
I resolved the failing test

tmeasday commented:
@ndelangen can we resolve [this question](https://github.com/storybookjs/storybook/pull/8765#issuecomment-553199004)? I'd like to understand what's going on.

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

# Conflicts:
#	addons/storyshots/storyshots-core/package.json

ndelangen committed:
CHANGE to use configure over copied loadStories fn

ndelangen committed:
ADD support for configure to accept an array of string pointing to files, without require.context

ndelangen committed:
CLEANUP && Use configure function from clientApi

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots

ndelangen committed:
ROLLBACK the change to allow for any string to be loadable, as this is nightmare-scenario for webpack
"Critical dependency: the request of a dependency is an expression"

Instead allow the array to contain "functions that do the require & return exports" or "the exports directly"

ndelangen commented:
@tmeasday @shilman WDYT now, better?

tmeasday commented on addons/storyshots/storyshots-core/src/frameworks/configure.ts:
```diff
@@ -1,27 +1,102 @@
+
+  if (stories && stories.length) {
+    // eslint-disable-next-line global-require, import/no-dynamic-require
+    storybook.configure(stories.map(f => require(f)), false);
```
Do we need to pass `framework` as the third argument to `configure()` (somehow)?

tmeasday replied:
```suggestion
    storybook.configure(() => stories.map(f => require(f)), false);
```

I think if you make this change you wouldn't need to change the definition of `configure`. @shilman is the expert though.

tmeasday commented on lib/core/src/client/preview/start.js:
```diff
@@ -350,13 +350,21 @@ export default function start(render, { decorateStory } = {}) {
-            typeof req.resolve === 'function' ? req.resolve(filename) : null
-          );
-        });
+        if (typeof req.keys === 'function') {
```
I don't think you should need to change this. @shilman and I already covered off this use case when defining this function I think.

tmeasday reviewed (changes requested):
Almost there I think!

ndelangen committed:
tom, you're correct as always!

ndelangen commented:
good to merge?

ndelangen committed:
Merge branch 'core/magic-entry' into core/magic-entry-storyshots
